### PR TITLE
[Editor] Implement the new alt text flow (bug 1909604)

### PR DESCRIPTION
**Pull request by calixteman**
For the Firefox pdf viewer, we want to use AI to guess an alt-text when adding an image to a pdf. For now the telemtry stuff is not implemented and will come soon.
In order to test it locally:
 - set enableAltText, enableFakeMLManager and enableUpdatedAddImage to true.
or in Firefox:
 - set browser.ml.enable, pdfjs.enableAltText and pdfjs.enableUpdatedAddImage to true.

### CODE REVIEW

File: l10n/en-US/viewer.ftl
Comment (flodolo) [inline l10n/en-US/viewer.ftl:451]:
```suggestion
    .aria-valuemax = { $totalSize }
```

Variables in Fluent are surrounded by `{ }`(also applies to all the other variables in these strings)

File: l10n/en-US/viewer.ftl
Comment (flodolo) [inline l10n/en-US/viewer.ftl:453]: Is it really bits and not bytes?

**Review by flodolo — CHANGES_REQUESTED** [review]

File: src/display/editor/alt_text.js
Comment (flodolo) [inline src/display/editor/alt_text.js:131]: This doesn't feel very safe. Can you use `data-l10n-attrs` to pass the variable?

Comment (calixteman) [reply src/display/editor/alt_text.js:131]: I did that because the generated string (from l10n engine) was containing some markers which are not in the ascii range.
So instead of workarounding, I fixed the bug:
https://github.com/mozilla/pdf.js/pull/18493

**Review by flodolo — APPROVED** [review]

**Comment by Snuffleupagus** [comment]
Given the sheer size of this PR, it'll take some time for me to properly review all of this. Leaving some high level observations and questions for now:
 - Since we have an upcoming release (maybe this weekend), this should *not* land just before that anyway.
 - There's been some changes/updates of the PR since it's opened, and I'd like to know that the patch has "stabilized" before doing a full review.
 - These changes mean that we now have *two separate* implementations for the altText-handling in the UI, what's the eventual plan for consolidating this functionality?
   Will the new dialog be usable in the GENERIC viewer, but obviously with AI generation switched off there, since it'd seem quite unfortunate if we'd need to maintain two separate implementations going forward?
 - The `FakeMLManager`-code should probably be limited, using the pre-processor, to just development- and testing-mode.
 - Looking at the preview above it no longer seems possible to mark an image as `decorative`, is that intended?
 - There seems to be a little bit more leading padding before the icon in the inline toolbar now, is this intended since it leads to more empty space there (see the arrow in the screen-shot below)?
    
   padding

**Comment by calixteman** [comment]
> Given the sheer size of this PR, it'll take some time for me to properly review all of this. Leaving some high level observations and questions for now:
> 
>     * Since we have an upcoming release (maybe this weekend), this should _not_ land just before that anyway.

The soft freeze in m-c is next week (thursday) and we really want it asap.
The integration tests are passing locally so hopefully nothing is broken.
That said if it causes a problem for the release, I'd prefer we postpone the release itself.

> 
>     * There's been some changes/updates of the PR since it's opened, and I'd like to know that the patch has "stabilized" before doing a full review.

It should be stabilized now.

> 
>     * These changes mean that we now have _two separate_ implementations for the altText-handling in the UI, what's the eventual plan for consolidating this functionality?

We want to experiment the new flow + the ai stuff in the next release and we'll decide what we want to do then.

>       Will the new dialog be usable in the GENERIC viewer, but obviously with AI generation switched off there, since it'd seem quite unfortunate if we'd need to maintain two separate implementations going forward?

Whatever we decide, I'll do the cleanup, but for now we must live with I guess. But for sure, we (/I) don't want to have a duplicated flow...

> 
>     * The `FakeMLManager`-code should probably be limited, using the pre-processor, to just development- and testing-mode.

Fixed.

> 
>     * Looking at the preview above it no longer seems possible to mark an image as `decorative`, is that intended?

Yes it's. 

> 
>     * Will the COMPONENTS-build work correctly with the new image-resources added here, note [this code](https://github.com/mozilla/pdf.js/blob/2efa3e460c94ad6ad48541e2ec44af6e22153bda/gulpfile.mjs#L1095-L1102)?

Fixed.

> 
>     * There seems to be a little bit more leading padding before the icon in the inline toolbar now, is this intended since it leads to more empty space there (see the arrow in the screen-shot below)?

Fixed.

**Comment by Snuffleupagus** [comment]
> The soft freeze in m-c is next week (thursday) and we really want it asap.

Well, I hope you understand that I cannot review well over a thousand lines of new code immediately (and I'm still not at a 100% unfortunately).
Especially since a lot of this is stuff is asynchronous, which requires more time to consider how the various pieces interact.

One quickly observed bug, related to the asynchronicity of the code:
 1. Open the preview in https://github.com/mozilla/pdf.js/pull/18492#issuecomment-2247947991.
 2. Add a new Stamp-editor.
 3. Immediately close the altText dialog *and* delete the editor.

In the console the following is printed (after a short wait):
```
TypeError: can't access property "guessedAltText", this[#currentEditor] is null
    #mlGuessAltText new_alt_text_manager.js:248
new_alt_text_manager.js:257:14
```

**Comment by calixteman** [comment]
> > The soft freeze in m-c is next week (thursday) and we really want it asap.
> 
> Well, I hope you understand that I cannot review well over a thousand lines of new code immediately (and I'm still not at a 100% unfortunately). Especially since a lot of this is stuff is asynchronous, which requires more time to consider how the various pieces interact.
>

I completely understand your point, but we want to make the experiment in 130.
If you see something obviously wrong I'll fix it immediately, but if you don't notice any bad things, we could merge it and fix the issues we may discover (especially when I'll begin to write tests or if you find them in playing with this new UI) in follow-up patches. Of course, if we discover something bad we can't fix easily, we won't ship it and we can back out it.
I'm a bit in a rush but I really understand you aren't and I don't want to rush you...
And I've few other things to implement... (telemetry and image settings mostly).
  
> One quickly observed bug, related to the asynchronicity of the code:
> 
>     1. Open the preview in [[Editor] Implement the new alt text flow (bug 1909604) #18492 (comment)](https://github.com/mozilla/pdf.js/pull/18492#issuecomment-2247947991).
> 
>     2. Add a new Stamp-editor.
> 
>     3. Immediately close the altText dialog _and_ delete the editor.
> 
> 
> In the console the following is printed (after a short wait):
> 
> ```
> TypeError: can't access property "guessedAltText", this[#currentEditor] is null
>     #mlGuessAltText new_alt_text_manager.js:248
> new_alt_text_manager.js:257:14
> ```

I fixed it two days ago.
I'll push a new version, with the enabled options in order to have a new viewer, and I'll set the options back, once we're more or less happy with patch.

**Comment by calixteman** [comment]
@Snuffleupagus and FYI, QA already started to play with the preview and reported to me some bugs they found and I fixed. They'll continue to test it.

**Comment by timvandermeij** [comment]
> Since we have an upcoming release (maybe this weekend), this should not land just before that anyway.

I'm going to make a new release this weekend, so that should be in time given that the soft-freeze for m-c is after the weekend. The advantage of releasing first is that we avoid any risks for the release because this is quite a big patch, and by merging this afterwards it'll have a month time to settle which gives us enough time to have all of this reviewed and any bugs we encounter afterwards fixed before the next release.

> Well, I hope you understand that I cannot review well over a thousand lines of new code immediately (and I'm still not at a 100% unfortunately).

Would it help if I do a pass over this PR during the weekend? I don't have extensive knowledge about the editor code, but perhaps an extra pair of eyes on the patch could help to increase confidence a bit if we need to land this in a short amount of time.

**Comment by calixteman** [comment]
> Would it help if I do a pass over this PR during the weekend? I don't have extensive knowledge about the editor code, but perhaps an extra pair of eyes on the patch could help to reduce the review burden and increase confidence a bit if we need to land this in a short amount of time.

Any help is helpful ! and appreciated, thank you.

File: src/display/editor/editor.js
Comment (timvandermeij) [inline src/display/editor/editor.js:1039]: It looks like `this.#altText` might be missing given the `#this.altText?` checks below, so shouldn't we also do that here and in the getter just above to make sure the access is safe?

File: web/firefoxcom.js
Comment (Snuffleupagus) [inline web/firefoxcom.js:355]: Should we not remove it from `this.#ready` as well, to avoid inconsistent state?

**Review by Snuffleupagus — COMMENTED** [review]
The implementation in `NewAltTextManager` feels a bit complicated, based on an initial look, but maybe it's not really possible to simplify it?

File: src/display/editor/editor.js
Comment (timvandermeij) [inline src/display/editor/editor.js:1047]: If `this.#altText`  is `null` this function returns `null` while I'd expect this to always return a boolean. Should we use `!!this.altText` instead?

File: web/app.js
Comment (timvandermeij) [inline web/app.js:209]: Tracking comment to make sure the commented out code here and below gets replaced before merging.

File: web/app_options.js
Comment (timvandermeij) [inline web/app_options.js:190]: Tracking comment to make sure the changes in this file get updated before merging (it's also what makes linting fail now).

File: web/genericcom.js
Comment (timvandermeij) [inline web/genericcom.js:63]: `guess(_data) {}` I assume, given that the `guess` method in `firefoxcom` has a `data` argument?

File: web/new_alt_text_manager.js
Comment (timvandermeij) [inline web/new_alt_text_manager.js:1]: s/2023/2024

File: web/new_alt_text_manager.js
Comment (timvandermeij) [inline web/new_alt_text_manager.js:35]: I would expect a variable starting with `is` to be a boolean instead of `null`.

File: web/new_alt_text_manager.js
Comment (timvandermeij) [inline web/new_alt_text_manager.js:90]: Should we use an `AbortController` to make sure that all event listeners get destroyed properly upon dialog close?

File: web/new_alt_text_manager.js
Comment (timvandermeij) [inline web/new_alt_text_manager.js:367]: Shouldn't we use `this.#uiManager?.setSelected` here given the statement two lines above?

File: web/viewer-geckoview.html
Comment (timvandermeij) [inline web/viewer-geckoview.html:68]: Let's move this down to keep the list ordered alphabetically for readability.

File: web/viewer.html
Comment (timvandermeij) [inline web/viewer.html:1]: We should revert this change. I assume it removed the UTF-8 BOM, but as mentioned in #18370 if we do that we should probably do it outside of this patch in case something breaks (we don't _expect_ it to, but it would be a shame if we'd have to fix it or back out this patch because of it).

**Review by timvandermeij — COMMENTED** [review]
Looks good to me overall. The comments below are everything I've found going over the entire patch, but I'd like to see the updated patch and play with the preview a bit more before approving this.

File: l10n/en-US/viewer.ftl
Comment (calixteman) [reply l10n/en-US/viewer.ftl:461]: It has been written by an American.

File: src/display/editor/editor.js
Comment (calixteman) [reply src/display/editor/editor.js:230]: Because the corresponding string has a parameter, so we need to call the got function with the parameter in order to get the translated string.

File: src/display/editor/editor.js
Comment (calixteman) [reply src/display/editor/editor.js:237]: Because the corresponding string has a parameter, so we need to call the got function with the parameter in order to get the translated string.

File: web/new_alt_text_manager.js
Comment (calixteman) [reply web/new_alt_text_manager.js:35]: Yep I agree, but it was to be able to have something like `if (this.#isEditing !== value)...` the first time we set it.

File: web/new_alt_text_manager.js
Comment (calixteman) [reply web/new_alt_text_manager.js:90]: This ctor is called only one time and set the different listeners for the different elements.
So removing them when closing the dialog implies to add them again when we reopen it: it isn't what we are doing in the current alt_text_manager.js and I followed the same pattern.
That said, I'm not sure to know what's the best approach. But for sure we can do that in a follow-up.

File: web/new_alt_text_manager.js
Comment (timvandermeij) [inline web/new_alt_text_manager.js:371]: If the PDF with image and alt text is saved to disk and opened again in PDF.js I don't see the alt text in the DOM. I would assume it to end up somewhere so screen readers can find it. Could you explain how that works?

**Review by timvandermeij — APPROVED** [review]
Looks good to me, with:

- the comments below and the two open tracking comments above addressed;
- passing CI and tests;
- a follow-up issue to create integration tests for the new alt text flow because currently we only test the current/old flow (note that we can wait with that until the new alt text flow is accepted and it's combined with removing the current/old flow).

File: web/new_alt_text_manager.js
Comment (calixteman) [reply web/new_alt_text_manager.js:375]: The alt-text is in the struct tree under the `canvas`:
image
That said maybe we should just add the alt-text as a aria-*** on the annotation itself.

**Comment by calixteman** [comment]
@Snuffleupagus is there anything blocking ?

**Review by Snuffleupagus — APPROVED** [review]

**Pull request merged**
